lib/charms/operator_libs_linux/v0/dnf.py

```diff
@@ -0,0 +1,462 @@
+                buffer.append(char)
+
+        result.append(
+            {
+                "name": ".".join(placeholder[:-1]),
+                "arch": ".".join(placeholder[-1:]),
+                "epoch": epoch,
+                "version": version,
+                "release": "".join(buffer),
+                "repo": pkg_repo[1:] if pkg_repo.startswith("@") else pkg_repo,
+                "state": state[cache_type],
+            }
+        )
+
+    return result
+
+
+class _Cache:
+    """Track the current state of the DNF package cache."""
+
+    def __new__(cls) -> "_Cache":
+        """Create new  _Cache object.
+
+        Returns:
+            (_Cache): New _Cache object.
+        """
+        if not hasattr(cls, f"{cls.__name__}__instance"):
+            cls.__instance = super(_Cache, cls).__new__(cls)
+            cls.__instance.__initialized = False
+        return cls.__instance
+
+    def __init__(self) -> None:
+        if self.__initialized:
+            return
+
+        self.__initialized = True
+        self.refresh()
+
+    @property
+    def installed(self) -> List[Dict[str, Union[str, _PackageState]]]:
+        """Get installed packages.
+
+        Returns:
+            (List[Dict[str, Union[str, _PackageState]]]):
+                Cache of installed packages.
+        """
+        return copy.deepcopy(self.__installed)
+
+    @property
+    def available(self) -> List[Dict[str, Union[str, _PackageState]]]:
+        """Get available packages.
+
+        Returns:
+            (List[Dict[str, Union[str, _PackageState]]]):
+                Cache of available packages.
+        """
+        return copy.deepcopy(self.__available)
+
+    def refresh(self) -> None:
+        """Refresh current package cache."""
+        with ThreadPoolExecutor() as executor:
+            installed = executor.submit(_load_cache, "installed")
+            available = executor.submit(_load_cache, "available")
+            self.__installed = installed.result()
+            self.__available = available.result()
+
+
+def _refresh(func: Callable) -> Callable:
+    """Refresh cache after function call.
+
+    Args:
+        func (Callable): Function to wrap.
+    """
+
+    @wraps(func)
+    def wrapper(*args, **kwargs) -> None:
+        func(*args, **kwargs)
+        _Cache().refresh()
+
+    return wrapper
+
+
+def _search(
+    cache: List[Dict[str, Union[str, _PackageState]]], name: str
+) -> Optional[Dict[str, Union[str, _PackageState]]]:
+    """Search DNF package cache.
+
+    Args:
+        cache (List[Dict[str, Union[str, _PackageState]]]):
+            DNF package cache to search.
+        name (str):
+            Name of package to locate.
+
+    Returns:
+        (Optional[Dict[str, Union[str, _PackageState]]]):
+            Located cache entry. Returns None if package is not found in cache.
+    """
+    for entry in cache:
+        if entry["name"] == name:
+            return entry
+
+    return None
+
+
+class _MetaDNF(type):
```

Comment:
Hmm, I am conflicted about this. I have this check here because I do not want users to be able to import the dnf charm library if their base does not have DNF installed. The idea behind this is that they will have an implementation like the following:

```python3
try:
    from charms.operator_libs_linux.v0.dnf import dnf
    pkg_mngr = dnf
except ImportError:
    import charms.operator_libs_linux.v0.apt as apt
    pkg_mngr = apt
```

That is why I check for dnf when the class is first defined at import time (hence the metaclass). As for the `dnf install ...`, I have that there because otherwise the `add_repo(...)` method will not work due to the `config-manager` dnf subcommand not being available.

Reply:
`_MetaDNF` was removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb